mesonbuild/linkers.py

```diff
@@ -633,7 +633,18 @@ class LLVMDynamicLinker(GnuLikeDynamicLinkerMixin, PosixDynamicLinkerMixin, Dyna
     This is only the posix-like linker.
     """
 
-    pass
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # Some targets don't seem to support this argument (windows, wasm, ...)
+        _, _, e = mesonlib.Popen_safe(self.exelist + self._apply_prefix('--allow-shlib-undefined'))
+        self.has_allow_shlib_undefined = not ('unknown argument: --allow-shlib-undefined' in e)
+
+    def get_allow_undefined_args(self) -> typing.List[str]:
+        if self.has_allow_shlib_undefined:
+            return self._apply_prefix('--allow-shlib-undefined')
+        return []
```

Comment:
looking at the LLVM code --allow-shlib-undefined has been implemented for ELF (so Linux, *BSD, etc) for a long time, but isn't implemented for PE/COFF. So we should do one of two things here:

1) Push the `has_link_argument` method into the DynamicLinker class and invoke that instead of opencoding it with Popen.
2) Add the MachineInfo class into the Linker like #5833 does and do somehting like:
```python
if self.info.is_windows():
    return []
return self._apply_prefix('--allow-shlib-undefined'
```

Reply:
I personally prefer the first approach. In general I think it's better to make less assumptions about available features.

Reply:
ping

Reply:
Yeah sorry, I got caught up in Taisei stuff. Now that we're done with a new release I'll go take a look into this now.

Reply:
Looks like this is basically implemented already; `DynamicLinker` has a `has_multi_arguments` method. Only problem is that it takes an `Environment` object, which I don't think is accessible in this context. I see a few other methods like `get_soname_args` and `build_rpath_args` take `Environment` as a parameter; would it be okay to do the same in `get_allow_undefined_args`? Is there any reason why we don't just instantiate linkers with a permanent reference to an `Environment`?

Reply:
> looking at the LLVM code --allow-shlib-undefined has been implemented for ELF (so Linux, *BSD, etc) for a long time, but isn't implemented for PE/COFF.

I think it cannot be implemented for PE/COFF shared libraries, as they cannot contain undefined symbols (every imported symbol is resolved to a providing library at link time).  

So any shared library which actually requires undefined symbols isn't going to be simply portable to PE/COFF (see test cases/common/121 shared module for an example of the required gymnastics)

However, since we don't mark shared libraries as having undefined symbols or not (c.f. `libtool -no-undefined`), I guess all we can do is not provide this flag where unsupported, and have the build fail on PE/COFF if it's actually needed.